Rpath test fails on major dev bigger 256 with assertion error

Trying:

```
PATH=$PWD/build/scripts-3.7:$PATH PYTHONPATH=$PWD/build/lib.linux-x86_64-3.7 python3 testing/rpathtest.py --failfast
.........Warning, making long socket did not fail
........................../usr/lib64/python3.7/unittest/case.py:628: ResourceWarning: unclosed file <_io.BufferedReader name=b'/home/elavarde/Public/rdiff-backup/build/testfiles/output/file.gz'>
  testMethod()
ResourceWarning: Enable tracemalloc to get the object allocation traceback
.testing/rpathtest.py:315: ResourceWarning: unclosed file <_io.BufferedWriter name=b'/home/elavarde/Public/rdiff-backup/build/testfiles/output/file.gz'>
  fp_out.write(s)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
....F
======================================================================
FAIL: testGetnums (__main__.FilenameOps)
Test getting file numbers
----------------------------------------------------------------------
Traceback (most recent call last):
  File "testing/rpathtest.py", line 282, in testGetnums
    assert devnums == (259, 0), devnums
AssertionError: (3, 0)

----------------------------------------------------------------------
Ran 41 tests in 1.115s

FAILED (failures=1)
```

My interpretation is that there is a modulo error in the code of `src/cmodule.c` when calculating major (3 is the result of 259 modulo 256, and no other device fails like this).

It needs to be added that `testing/ctest.py` shows the same behaviour if I add `/dev/nvme0n1` to the list of files checked. At the same time, it shows that the python version of `make_file_dict` works correctly.

I see two options:

1. get rid of cmodule.c and replace with the python version for all platforms - that's probably the best option long term but I'm not sure about the performance impact.
2. fix cmodule.c - it probably means importing /usr/include/sys/sysmacros.h and/or /usr/include/bits/sysmacros.h instead of defining our own version of `major`.

Could someone with more C-knowledge have a look at the code and make a suggestion?

--- a/src/cmodule.c
+++ b/src/cmodule.c
@@ -30,186 +30,16 @@
 #include <errno.h>
 
 
-/* Some of the following code to define major/minor taken from code by
- * Jörg Schilling's star archiver.
- */
-#if !defined(major) && (defined(sgi) || defined(__sgi) || defined(__SVR4)) && !defined(__CYGWIN32__)
-#include <sys/mkdev.h>
-#endif
-
-#ifndef major
-#	define major(dev)		(((dev) >> 8) & 0xFF)
-#	define minor(dev)		((dev) & 0xFF)
-#	define makedev(majo, mino)	(((majo) << 8) | (mino))
-#endif
-/* End major/minor section */
-
 /* choose the appropriate stat and fstat functions and return structs */
 /* This code taken from Python's posixmodule.c */
-#undef STAT
 #if defined(MS_WIN64) || defined(MS_WIN32)
 #	define SYNC _flushall
 #else
-#	define LSTAT lstat
-#	define STAT stat
-#	define FSTAT fstat
-#	define STRUCT_STAT struct stat
 #	define SYNC sync
-#endif
-#ifndef PY_LONG_LONG 
-    #define PY_LONG_LONG LONG_LONG 
-#endif
-
-/* The following section is by Jeffrey A. Marshall and compensates for
- * a bug in Mac OS X's S_ISFIFO and S_ISSOCK macros.
- * Note: Starting in Mac OS X 10.3, the buggy macros were changed to be
- * the same as the ones below.
- */
-#ifdef __APPLE__
-/* S_ISFIFO/S_ISSOCK macros from <sys/stat.h> on mac osx are bogus */
-#undef S_ISSOCK               /* their definition of a socket includes fifos */
-#undef S_ISFIFO               /* their definition of a fifo includes sockets */
-#define S_ISSOCK(mode)        (((mode) & S_IFMT) == S_IFSOCK)
-#define S_ISFIFO(mode)        (((mode) & S_IFMT) == S_IFIFO)
 #endif
 
 static PyObject *UnknownFileTypeError;
-static PyObject *c_make_file_dict(PyObject *self, PyObject *args);
 static PyObject *my_sync(PyObject *self, PyObject *args);
-
-/* Turn a stat structure into a python dictionary.  The preprocessor
-   stuff taken from Python's posixmodule.c */
-static PyObject *c_make_file_dict(self, args)
-	 PyObject *self;
-	 PyObject *args;
-{
-#if defined(MS_WINDOWS)
-	PyErr_SetString(PyExc_AttributeError, "This function is not implemented on Windows.");
-	return NULL;
-#else
-  PyObject *size, *inode, *mtime, *atime, *ctime, *devloc, *return_val;
-  char *filename, filetype[5];
-  STRUCT_STAT sbuf;
-  long int mode, perms;
-  int res;
-
-  if (!PyArg_ParseTuple(args, "y", &filename)) return NULL;
-
-  Py_BEGIN_ALLOW_THREADS
-  res = LSTAT(filename, &sbuf);
-  Py_END_ALLOW_THREADS
-
-  if (res != 0) {
-	if (errno == ENOENT || errno == ENOTDIR)
-	  return Py_BuildValue("{s:s}", "type", NULL);
-	else {
-	  PyErr_SetFromErrnoWithFilename(PyExc_OSError, filename);
-	  return NULL;
-	}
-  }
-
-#ifdef HAVE_LARGEFILE_SUPPORT
-  size = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_size);
-  inode = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_ino);
-#else
-  size = PyLong_FromLong(sbuf.st_size);
-  inode = PyLong_FromLong((long)sbuf.st_ino);
-#endif /* HAVE_LARGEFILE_SUPPORT */
-  mode = (long)sbuf.st_mode;
-  perms = mode & 07777;
-#if defined(HAVE_LONG_LONG)
-  devloc = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_dev);
-#else
-  devloc = PyLong_FromLong((long)sbuf.st_dev);
-#endif
-#if SIZEOF_TIME_T > SIZEOF_LONG
-  mtime = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_mtime);
-  atime = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_atime);
-  ctime = PyLong_FromLongLong((PY_LONG_LONG)sbuf.st_ctime);
-#else
-  mtime = PyLong_FromLong((long)sbuf.st_mtime);
-  atime = PyLong_FromLong((long)sbuf.st_atime);
-  ctime = PyLong_FromLong((long)sbuf.st_ctime);
-#endif
-
-  /* Build return dictionary from stat struct */
-  if (S_ISREG(mode) || S_ISDIR(mode) || S_ISSOCK(mode) || S_ISFIFO(mode)) {
-	/* Regular files, directories, sockets, and fifos */
-	if S_ISREG(mode) strcpy(filetype, "reg");
-	else if S_ISDIR(mode) strcpy(filetype, "dir");
-	else if S_ISSOCK(mode) strcpy(filetype, "sock");
-	else strcpy(filetype, "fifo");
-	return_val =  Py_BuildValue("{s:s,s:O,s:l,s:l,s:l,s:O,s:O,s:l,s:O,s:O,s:O}",
-								"type", filetype,
-								"size", size,
-								"perms", perms,
-								"uid", (long)sbuf.st_uid,
-								"gid", (long)sbuf.st_gid,
-								"inode", inode,
-								"devloc", devloc,
-								"nlink", (long)sbuf.st_nlink,
-								"mtime", mtime,
-								"atime", atime,
-								"ctime", ctime);
-  } else if S_ISLNK(mode) {
-	/* Symbolic links */
-	char linkname[1024];
-	int len_link = readlink(filename, linkname, 1023);
-	if (len_link < 0) {
-	  PyErr_SetFromErrno(PyExc_OSError);
-	  return_val = NULL;
-	} else {
-	  linkname[len_link] = '\0';
-	  return_val = Py_BuildValue("{s:s,s:O,s:l,s:l,s:l,s:O,s:O,s:l,s:y}",
-								 "type", "sym",
-								 "size", size,
-								 "perms", perms,
-								 "uid", (long)sbuf.st_uid,
-								 "gid", (long)sbuf.st_gid,
-								 "inode", inode,
-								 "devloc", devloc,
-								 "nlink", (long)sbuf.st_nlink,
-								 "linkname", linkname);
-	}
-  } else if (S_ISCHR(mode) || S_ISBLK(mode)) {
-	/* Device files */
-	char devtype[2];
-#if defined(HAVE_LONG_LONG)
-	PY_LONG_LONG devnums = (PY_LONG_LONG)sbuf.st_rdev;
-	PyObject *major_num = PyLong_FromLongLong(major(devnums));
-#else
-	long int devnums = (long)sbuf.st_dev;
-	PyObject *major_num = PyLong_FromLong(major(devnums));
-#endif
-	int minor_num = (int)(minor(devnums));
-	if S_ISCHR(mode) strcpy(devtype, "c");
-	else strcpy(devtype, "b");
-	return_val = Py_BuildValue("{s:s,s:O,s:l,s:l,s:l,s:O,s:O,s:l,s:N}",
-							   "type", "dev",
-							   "size", size,
-							   "perms", perms,
-							   "uid", (long)sbuf.st_uid,
-							   "gid", (long)sbuf.st_gid,
-							   "inode", inode,
-							   "devloc", devloc,
-							   "nlink", (long)sbuf.st_nlink,
-							   "devnums", Py_BuildValue("(s,O,i)", devtype,
-														major_num, minor_num));
-	Py_DECREF(major_num);
-  } else {
-	/* Unrecognized file type - raise exception */
-	PyErr_SetString(UnknownFileTypeError, filename);
-	return_val = NULL;
-  }
-  Py_DECREF(size);
-  Py_DECREF(inode);
-  Py_DECREF(devloc);
-  Py_DECREF(mtime);
-  Py_DECREF(atime);
-  Py_DECREF(ctime);
-  return return_val;
-#endif /* defined(MS_WINDOWS) */
-}
 
 
 /* Run sync() and return None */
@@ -347,8 +177,6 @@ static PyObject *acl_unquote(PyObject *self, PyObject *args)
 /* ------------- Python export lists -------------------------------- */
 
 static PyMethodDef CMethods[] = {
-  {"make_file_dict", c_make_file_dict, METH_VARARGS,
-   "Make dictionary from file stat"},
   {"sync", my_sync, METH_VARARGS, "sync buffers to disk"},
   {"acl_quote", acl_quote, METH_VARARGS,
    "Quote string, escaping non-printables"},

--- a/src/rdiff_backup/rpath.py
+++ b/src/rdiff_backup/rpath.py
@@ -282,29 +282,10 @@ def make_file_dict(filename):
 	filename over the network, thus avoiding the need to pickle an
 	(incomplete) rpath object.
 	"""
-	if os.name != 'nt':
-		try:
-			return C.make_file_dict(filename)
-		except OSError as error:
-			# Unicode filenames should be processed by the Python version
-			if error.errno != errno.EILSEQ and error.errno != errno.EINVAL:
-				raise
-		except UnicodeEncodeError as error:
-			pass
 
-	return make_file_dict_python(filename)
-
-def make_file_dict_python(filename):
-	"""Create the data dictionary using a Python call to os.lstat
-	
-	We do this on Windows since Python's implementation is much better
-	than the one in cmodule.c    Eventually, we will move to using
-	this on all platforms since CPUs have gotten much faster than
-	they were when it was necessary to write cmodule.c
-	"""
 	try:
 		statblock = os.lstat(filename)
-	except os.error:
+	except (FileNotFoundError, NotADirectoryError):
 		return {'type':None}
 	data = {}
 	mode = statblock[stat.ST_MODE]

--- a/testing/ctest.py
+++ b/testing/ctest.py
@@ -5,22 +5,6 @@ from rdiff_backup.rpath import *
 
 class CTest(unittest.TestCase):
 	"""Test the C module by comparing results to python functions"""
-	def test_make_dict(self):
-		"""Test making stat dictionaries"""
-		rp1 = RPath(Globals.local_connection, "/dev/ttyS1")
-		rp2 = RPath(Globals.local_connection, "./ctest.py")
-		rp3 = RPath(Globals.local_connection, "aestu/aeutoheu/oeu")
-		rp4 = RPath(Globals.local_connection, "testfiles/various_file_types/symbolic_link")
-		rp5 = RPath(Globals.local_connection, "testfiles/various_file_types/fifo")
-
-		for rp in [rp1, rp2, rp3, rp4, rp5]:
-			dict1 = make_file_dict_python(rp.path)
-			dict2 = C.make_file_dict(rp.path)
-			if dict1 != dict2:
-				print("Python dictionary: ", dict1)
-				print("not equal to C dictionary: ", dict2)
-				print("for path ", rp.path)
-				assert 0
 
 	def test_sync(self):
 		"""Test running C.sync"""

